streams/src/main/java/org/apache/kafka/streams/state/internals/ReadonlyPartiallyDeserializedSegmentValue.java

```diff
@@ -0,0 +1,210 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.streams.query.ResultOrder;
+
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.Map;
+
+
+final class ReadonlyPartiallyDeserializedSegmentValue {
+
+    private static final int TIMESTAMP_SIZE = 8;
+    private static final int VALUE_SIZE = 4;
+    private byte[] segmentValue;
+    private long nextTimestamp;
+    private long minTimestamp;
+
+    private int deserIndex = -1; // index up through which this segment has been deserialized (inclusive)
+
+    private Map<Integer, Integer> cumulativeValueSizes;
+
+    private int valuesStartingIndex = -1; // the index of the first value in the segment (but the last one in the list)
+    private Map<Integer, TimestampAndValueSize> unpackedTimestampAndValueSizes = new HashMap<>();
+    private int recordNumber = -1; // number of segment records
+
+
+    ReadonlyPartiallyDeserializedSegmentValue(final byte[] segmentValue) {
+        this.segmentValue = segmentValue;
+        this.nextTimestamp =
+                RocksDBVersionedStoreSegmentValueFormatter.getNextTimestamp(segmentValue);
+        this.minTimestamp =
+                RocksDBVersionedStoreSegmentValueFormatter.getMinTimestamp(segmentValue);
+        resetDeserHelpers();
+    }
+
+
+    public long minTimestamp() {
+        return minTimestamp;
+    }
+
+    public long nextTimestamp() {
+        return nextTimestamp;
+    }
+
+    public byte[] serialize() {
+        return segmentValue;
+    }
+
+
+    public RocksDBVersionedStoreSegmentValueFormatter.SegmentValue.SegmentSearchResult find(
+            final long fromTime, final long toTime, final ResultOrder order, final int index) {
+
+        // this segment does not have any record in query specified time range
+        if (toTime < minTimestamp || fromTime > nextTimestamp) {
+            return null;
+        }
+
+        final boolean isAscending = order.equals(ResultOrder.ASCENDING);
+
+        if (isAscending && valuesStartingIndex == -1) {
+            findValuesStartingIndex();
+            deserIndex = recordNumber;
+        }
+
+        long currTimestamp = -1;
+        long currNextTimestamp = -1;
+        int currIndex = initializeCurrentIndex(index, isAscending);
+        int cumValueSize = initializeCumValueSize(index, currIndex, isAscending);
+        int currValueSize;
+
+
+        while (hasStillRecord(currTimestamp, currNextTimestamp, order)) {
+            if (hasBeenDeserialized(isAscending, currIndex)) {
+                final TimestampAndValueSize curr;
+                curr = unpackedTimestampAndValueSizes.get(currIndex);
+                currTimestamp = curr.timestamp;
+                cumValueSize = cumulativeValueSizes.get(currIndex);
```

Comment:
nit: flip order (first use `curr` to get out both field -- breaks the flow reading the code)